packages/protocol/src/protocol-thp/crypto/crypto.ts

```diff
@@ -0,0 +1,219 @@
+    tag: Buffer,
+) => {
+    const aesGcm = new AESGCM(key, getIvFromNonce(recvNonce));
+    // aesGcm.auth(handshakeHash);
+    aesGcm.auth(Buffer.alloc(0));
+    const trezorMaskedStaticPubkey = aesGcm.decrypt(payload, tag);
+
+    return trezorMaskedStaticPubkey.subarray(1); // session_id
+    // return trezorMaskedStaticPubkey;
+};
+
+export const handleCreateChannelResponse = (deviceProperties: Buffer) => {
+    // 1. Set h = SHA-256(protocol_name || device_properties).
+    return hashOfTwo(PROTOCOL_NAME, deviceProperties);
+};
+
+// called after handshakeCompletionRequest
+export const updateKeys = (credentials: ThpHandshakeCredentials, payload: Buffer) => {
+    const handshakeHash = hashOfTwo(credentials.handshakeHash, payload);
+    const [hostKey, trezorKey] = hkdf(credentials.trezorKey, Buffer.alloc(0));
+
+    // 2. Set trezor_state, success = AES-GCM-DECRYPT(key=key_response, IV=0^96, ad=empty_string, plaintext=trezor_state). Assert that success is True.
+    // const aesGcm = new AESGCM(hostKey, Buffer.alloc(12));
+    // aesGcm.auth(handshakeHash);
+    // const trezorState =
+
+    return {
+        ...credentials,
+        handshakeHash,
+        trezorKey,
+        hostKey,
+    };
+};
+
+// 10. Search credentials for a pairs (trezor_static_pubkey, credential) such that trezor_masked_static_pubkey == X25519(SHA-256(trezor_static_pubkey || trezor_ephemeral_pubkey), trezor_static_pubkey).
+export const findKnownPairingCredentials = (
+    trezorStaticPubkey: Buffer,
+    trezorEphemeralPubkey: Buffer,
+) => {
+    // X25519(SHA-256(trezor_static_pubkey || trezor_ephemeral_pubkey), trezor_static_pubkey).
+    const h = hashOfTwo(trezorStaticPubkey, trezorEphemeralPubkey);
+
+    return curve25519(h, trezorStaticPubkey);
+};
+
+export const getTrezorState = (credentials: ThpHandshakeCredentials, payload: Buffer) => {
+    // 2. Set trezor_state, success = AES-GCM-DECRYPT(key=key_response, IV=0^96, ad=empty_string, plaintext=trezor_state). Assert that success is True.
+    const aesGcm = new AESGCM(credentials.trezorKey, Buffer.alloc(12));
+    aesGcm.auth(Buffer.alloc(0));
+    const trezorState = aesGcm.decrypt(payload.subarray(0, 1), payload.subarray(1, 17));
+
+    return trezorState.readUint8();
+};
+
+type ThpCryptoOptions = {
+    handshakeHash: Buffer;
+    trezorEphemeralPubkey: Buffer;
+    trezorEncryptedStaticPubkey: Buffer;
+    tag: Buffer;
+    sendNonce: number;
+    recvNonce: number;
+    // knownCredentials: any[];
+};
+
+export const handleHandshakeInitResponse = ({
+    handshakeHash,
+    trezorEphemeralPubkey,
+    trezorEncryptedStaticPubkey,
+    tag,
+    // knownCredentials,
+    sendNonce,
+    recvNonce,
+}: ThpCryptoOptions) => {
+    const iv0 = getIvFromNonce(sendNonce); // should be 0
+    const iv1 = getIvFromNonce(recvNonce); // should be 1
+
+    const hostStaticKeys = getCurve25519KeyPair(Buffer.alloc(32).fill(9));
+    const hostEphemeralKeys = getCurve25519KeyPair(Buffer.alloc(32).fill(1));
```

Comment:
This looks suspicious. The host ephemeral key pair should be generated randomly before the `HandshakeInitiationReq` is sent, since its public part is contained in the message.